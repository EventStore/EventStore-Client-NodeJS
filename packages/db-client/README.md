# @eventstore/db-client

[![npm][npm-badge]][npm-badge-url]
[![Github action CI workflow][ci-badge]][ci-badge-url]
[![license][license-badge]][license-badge-url]

This is the package for the NodeJS client for EventStoreDB 20+ and uses gRPC as the communication protocol.

## Support

Information on support can be found on our website: [Event Store Support]

## Communities

- [Discuss]
- [Discord (Event Store)][discord-event-store]
- [Discord (ddd-cqrs-es)][Discord-ddd-cqrs-es]

## Contributing

Refer to our [contribution guidelines][contributing-guidelines].

[event store support]: https://eventstore.com/support/
[discuss]: https://discuss.eventstore.com/
[discord-event-store]: https://discord.gg/Phn9pmCw3t
[Discord-ddd-cqrs-es]: https://discord.com/invite/sEZGSHNNbH
[contributing-guidelines]: https://github.com/prisma/prisma/blob/main/CONTRIBUTING.md
[npm-badge]: https://img.shields.io/npm/v/@eventstore/db-client.svg
[npm-badge-url]: https://www.npmjs.com/package/@eventstore/db-client
[ci-badge]: https://github.com/EventStore/EventStore-Client-NodeJS/workflows/CI/badge.svg?branch=master
[ci-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/actions
[license-badge]: https://img.shields.io/npm/l/@eventstore/db-client.svg
[license-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/LICENSE