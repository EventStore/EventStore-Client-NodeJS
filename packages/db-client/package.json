{
  "name": "@eventstore/db-client",
  "version": "6.2.1",
  "description": "NodeJS EventStoreDB version 20+ and uses gRPC as the communication protocol.",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "npm-run-all generate build:ts",
    "build:ts": "tsc --build",
    "build:watch": "npm-run-all generate build:ts-watch",
    "build:ts-watch": "tsc --build --watch",
    "generate": "run-s generate:*",
    "generate:sed-append": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*) \\[deprecated = true\\];/int64 \\$1 = \\$2 [deprecated = true, jstype = JS_STRING];/g\" ./protos/*",
    "generate:sed-insert": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*);/int64 \\$1 = \\$2 [jstype = JS_STRING];/g\" ./protos/*",
    "generate:folder": "shx mkdir -p ./generated",
    "generate:js": "grpc_tools_node_protoc ./protos/*.proto --proto_path=./protos --grpc_out=grpc_js:./generated --js_out=import_style=commonjs,binary:./generated --ts_out=generate_package_definition:./generated",
    "clean": "run-s clean:*",
    "clean:generated": "shx rm -rf ./generated",
    "clean:dist": "shx rm -rf ./dist",
    "clean:buildinfo": "shx rm -rf *.tsbuildinfo"
  },
  "files": [
    "dist/",
    "generated/"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/EventStore/EventStore-Client-NodeJS.git"
  },
  "keywords": [
    "EventStore",
    "EventSourcing",
    "Client"
  ],
  "author": "Event Store Limited",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/EventStore/EventStore-Client-NodeJS/issues"
  },
  "homepage": "https://developers.eventstore.com/clients/grpc/getting-started?codeLanguage=NodeJS",
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": "^12.18.3 || >=14"
  },
  "dependencies": {
    "@grpc/grpc-js": "^1.9.12",
    "@types/debug": "^4.1.12",
    "@types/google-protobuf": "^3.15.12",
    "@types/node": "^16.18.67",
    "debug": "^4.3.2",
    "google-protobuf": "^3.21.2",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.4",
    "grpc-tools": "^1.12.4",
    "grpc_tools_node_protoc_ts": "^5.3.3",
    "npm-run-all": "^4.1.5",
    "shx": "^0.3.4"
  }
}
