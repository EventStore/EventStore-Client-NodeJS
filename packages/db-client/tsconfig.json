{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "module": "node16",
    "outDir": "./dist",
    "rootDir": "./src"
  },
  "include": [
    "./src/**/*.ts",
    "./generated"
  ],
  "references": [
    {
      "path": "../db-client-bridge"
    }
  ]
}
