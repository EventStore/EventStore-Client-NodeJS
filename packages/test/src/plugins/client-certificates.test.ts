/* eslint-disable @typescript-eslint/no-explicit-any */
import { createTestNode, jsonTestEvents } from "@test-utils";
import { AccessDeniedError, KurrentDBClient } from "@kurrent/db-client";

describe("client certificates", () => {
  const node = createTestNode();

  beforeAll(async () => {
    await node.up();
  });

  afterAll(async () => {
    await node.down();
  });

  describe("throw error when constructor is initialised with incorrect channel credentials combinations", () => {
    test.each([
      [
        "userCertFile",
        () =>
          new KurrentDBClient(
            { endpoint: node.uri },
            {
              rootCertificate: node.certs.root,
              userCertFile: node.certs.users.admin.userCertFile,
            }
          ),
      ],
      [
        "userKeyFile",
        () =>
          new KurrentDBClient(
            { endpoint: node.uri },
            {
              rootCertificate: node.certs.root,
              userKeyFile: node.certs.users.admin.userKeyFile,
            }
          ),
      ],
    ])("constructor initialised with %s only", (_, makeCall) => {
      try {
        makeCall();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });

    test.each([
      [
        "userCertFile",
        () =>
          KurrentDBClient.connectionString`esdb://${node.uri}?tls=true&tlsCAFile=${node.certPath.root}&userCertFile=${node.certPath.admin.certPath}`,
      ],
      [
        "userKeyFile",
        () =>
          KurrentDBClient.connectionString`esdb://${node.uri}?tls=true&tlsCAFile=${node.certPath.root}&userKeyFile=${node.certPath.admin.certKeyPath}`,
      ],
    ])("connection string with %s only", (_, connection) => {
      try {
        connection();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  describe("client initialized with only the admin certificate", () => {
    let client: KurrentDBClient;

    beforeEach(() => {
      client = new KurrentDBClient(
        { endpoint: node.uri },
        {
          rootCertificate: node.certs.root,
          userCertFile: node.certs.users.admin.userCertFile,
          userKeyFile: node.certs.users.admin.userKeyFile,
        }
      );
    });

    test("using default only certificates", async () => {
      const result = await client.appendToStream(
        "using_default_only_certificates",
        jsonTestEvents(2)
      );
      expect(result).toBeDefined();
    });

    test("overridden user credentials takes precedence over invalid user certificate", async () => {
      await expect(
        client.appendToStream(
          "overridden_user_credentials_takes_precedence",
          jsonTestEvents(2),
          {
            credentials: {
              username: "wrong",
              password: "password",
            },
          }
        )
      ).rejects.toThrow(AccessDeniedError);
    });
  });

  test("user credentials takes precedence over the client certificate during initialization", async () => {
    const clientWithCredentials = new KurrentDBClient(
      { endpoint: node.uri },
      {
        rootCertificate: node.certs.root,
        userCertFile: node.certs.users.admin.userCertFile,
        userKeyFile: node.certs.users.admin.userKeyFile,
      },
      {
        username: "wrong",
        password: "password",
      }
    );

    await expect(
      clientWithCredentials.appendToStream(
        "user_credentials_takes_precedence_over_client_certificate",
        jsonTestEvents(2)
      )
    ).rejects.toThrow(AccessDeniedError);
  });

  test("When the client is initialized with invalid certificate, user credentials take precendence if overriden during a call", async () => {
    const clientWithBadCertificate = new KurrentDBClient(
      { endpoint: node.uri },
      {
        rootCertificate: node.certs.root,
        userCertFile: node.certs.users.invalid.userCertFile,
        userKeyFile: node.certs.users.invalid.userKeyFile,
      }
    );

    expect(
      await clientWithBadCertificate.appendToStream(
        "user_credentials_takes_precedence_over_client_certificate",
        jsonTestEvents(2),
        {
          credentials: {
            username: "admin",
            password: "changeit",
          },
        }
      )
    ).toBeDefined();
  });
});
