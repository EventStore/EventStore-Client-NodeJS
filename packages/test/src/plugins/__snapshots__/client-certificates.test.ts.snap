// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations connection string with userCertFile only 1`] = `[Error: userCertFile must be given with accompanying userKeyFile]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations connection string with userKeyFile only 1`] = `[Error: userCertFile must be given with accompanying userKeyFile]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations constructor initialised with userCertFile only 1`] = `[Error: Certificate chain must be given with accompanying private key]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations constructor initialised with userKeyFile only 1`] = `[Error: Private key must be given with accompanying certificate chain]`;
